agent/pkg/cruntimes/crutils.go

```diff
@@ -0,0 +1,230 @@
+
+		if err := p.Publish(ctx, docker.NewTypedLogEvent(level, log, stdtype)); err != nil {
+			logrus.WithError(err).Trace("log stream terminated")
+			return
+		}
+	}
+}
+
+// PprintCommand provides pretty printing of the given command to trace & log outputs.
+func PprintCommand(
+	ctx context.Context,
+	command string,
+	args []string,
+	p events.Publisher[docker.Event],
+	log *logrus.Entry,
+) error {
+	toPrint := command
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "--") { // print each arg on a new line
+			toPrint += " \\\n"
+			toPrint += "\t"
+			toPrint += arg
+		} else {
+			toPrint += " "
+			toPrint += arg
+		}
+	}
+
+	log.Trace(toPrint)
+	if err := p.Publish(ctx, docker.NewLogEvent(
+		model.LogLevelDebug,
+		toPrint,
+	)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CanonicalizeImage returns the canonicalized image name.
+func CanonicalizeImage(image string) string {
+	url, err := url.Parse(image)
+	isURIForm := err == nil
+	isFSForm := path.IsAbs(image)
+	if isFSForm || (isURIForm && url.Scheme != "") {
+		return image
+	}
+	return fmt.Sprintf("docker://%s", image)
+}
+
+// PullImage implements code sharing for singularity & podman.
+func PullImage(
+	ctx context.Context,
+	req docker.PullImage,
+	p events.Publisher[docker.Event],
+	wg *waitgroupx.Group,
+	log *logrus.Entry,
+	getPullCommand func(docker.PullImage, string) (string, []string),
+) (err error) {
+	if err = p.Publish(ctx, docker.NewBeginStatsEvent(docker.ImagePullStatsKind)); err != nil {
+		return err
+	}
+	defer func() {
+		if err = p.Publish(ctx, docker.NewEndStatsEvent(docker.ImagePullStatsKind)); err != nil {
+			log.WithError(err).Warn("did not send image pull done stats")
+		}
+	}()
+
+	image := CanonicalizeImage(req.Name)
+
+	uri, err := url.Parse(image)
+	if err != nil || uri.Scheme == "" {
+		if err = p.Publish(ctx, docker.NewLogEvent(
+			model.LogLevelInfo,
+			fmt.Sprintf("image %s isn't a pullable URI; skipping pull", image),
+		)); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// TODO(DET-9078): Support registry auth. Investigate other auth mechanisms
+	// with singularity & podman.
+	command, args := getPullCommand(req, image)
+
+	if err = PprintCommand(ctx, command, args, p, log); err != nil {
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, command, args...) // #nosec G204 'command' is under our control
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("creating stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("creating stderr pipe: %w", err)
+	}
+
+	// The return codes from `podman pull` aren't super helpful in determining the error, so we
+	// wrap the publisher and skim logs to see what happened as we ship them.
+	ignoreErrorsSig := make(chan bool)
+	checkIgnoreErrors := events.FuncPublisher[docker.Event](
+		func(ctx context.Context, t docker.Event) error {
+			if t.Log != nil && strings.Contains(t.Log.Message, "Image file already exists") {
```

Comment:
Singularity, Podman, and Enroot all issue the message "Image file already exists"?